### ISSUE TITLE
The multi_gpu_model problem

### ISSUE THREAD

**Issue by FlyEgle** [qqwweee/keras-yolo3#151]
The yolo last layers is loss layers , the output is a shape(1, ) tensor. So if directly use the multi_gpu_model, it could have wrong , beacause the concat is not receive the length of list of output is 1. So you should overite the multi_gpu_model ,and change concate to add for loss tensor calculate, then have the mean loss

**Comment by pzxdd** [comment]
could you please share your code about overwrite the multi_gpu_model?
thx!

**Comment by zxwxz** [comment]
Is directly replace concatenate into add has same logical meaning?(suppose training output only loss value)
I modify a new .py to replace keras/utils/multi_gpu_utils.py
multi_gpu_model can start without error
`
with tf.device('/cpu:0' if cpu_merge else '/gpu:%d' % target_gpu_ids[0]):
        merged = []
        for name, outputs in zip(model.output_names, all_outputs):
            merged.append(concatenate(outputs,
                                      axis=0, name=name))
`

**Comment by tanakataiki** [comment]
@FlyEgle 
Well, addition insted of concat doesnt detect anything. 
Did you have better or same result using multigpu training script?
The loss didnt decrease from 30.
The parameters are the same as you.

**Comment by tanakataiki** [comment]
@FlyEgle 
Thanks Ive tested for 2 days training with your code though .

**Comment by zxwxz** [comment]
@FlyEgle @tanakataiki 
I think training speed is not cause by GPU numbers
After analysis, data_generator_wrapper will block during training
I apply a below change can speed up 3 times
`def data_generator(annotation_lines, batch_size, input_shape, anchors, num_classes):
    '''data generator for fit_generator'''
    n = len(annotation_lines)
    i = 0
    while True:
        image_data = []
        box_data = []
        if (i+batch_size > n):
            np.random.shuffle(annotation_lines)
            i = 0
        output = threadpool.starmap(get_random_data, zip(annotation_lines[i:i+batch_size], itertools.repeat(input_shape, batch_size)))
        image_data = list(zip(*output))[0]
        box_data = list(zip(*output))[1]
        i = i+batch_size
        image_data = np.array(image_data)
        box_data = np.array(box_data)
        y_true = preprocess_true_boxes(box_data, input_shape, anchors, num_classes)
        yield [image_data, *y_true], np.zeros(batch_size)`
I also separate data generate/consume into different thread, but not really help a lot according to cpu compute bound

**Comment by tanakataiki** [comment]
@zxwxz 
The Thredpool  seems not to work on windows.
and I also met the hypothesis that the number of gpu doesnt affect speeding up.
could you write your code with indention?

**Comment by zxwxz** [comment]
I update my change here
https://github.com/zxwxz/keras-yolo3/commit/51d55bbe4c680a7f8ae2bf995dd4929809bb77dc
I think data feeding should be the critical session, original darknet implementation might be a good reference, it can process over 3000 images in 1 sec during training.
Another problem might be in get_random_data, random flag set true cause the loss cannot under 10
After disable random, can continued decrease, but still mush slower compare to darknet(training speed problem)

**Comment by mazatov** [comment]
Have you guys figured this out? 

I have the same issue @tanakataiki .  I implemented the multi_gpu_model function written by @FlyEgle and I have no improvement on time.

**Comment by wangzilu** [comment]
@zxwxz Hello, I tested your code, actually not correctly data reading.

**Comment by AlbertoSabater** [comment]
Anyone has achieved training time improvement?
I'm working with multi_gpu_model and Image Augmentation in parallel as @zxwxz does, but the training time is even worse.
Looks like the bottleneck is the CPU that performs de Image Augmentation, so adding more GPUs doesn't improve the performance.

**Comment by anvien** [comment]
https://github.com/qqwweee/keras-yolo3/issues?page=11&q=is%3Aissue+is%3Aopen

**Comment by zhanganguo** [comment]
I have pushed the useable code for training multi_gpu_model for YOLOv3 with multiple backbones, please visit https://github.com/anvien/Multi-YOLOv3

**Comment by BITOPYX** [comment]
Thank you very much! Your code really really help me a lot!
